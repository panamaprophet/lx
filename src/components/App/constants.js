export const DATA_URL = '//www.mocky.io/v2/5d3fec2b33000062009d27bc';

export const DEFAULT_ACTIVE_MENU_ITEM = -1;