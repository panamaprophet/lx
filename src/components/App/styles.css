.root {
	font-family: sans-serif;
    font-size: 12px;
    user-select: none;
}