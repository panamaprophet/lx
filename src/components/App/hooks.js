import {useEffect} from 'react';

import {DEFAULT_ACTIVE_MENU_ITEM} from './constants';

/**
 * normalize data for some consistency between items and subitems
 */
const normalizeData = data => data.map(item => {
    const {subItems, ...rest} = item;

    return {
        ...rest, 
        items: subItems,
    };
});

export const useOutsideClickEffect = 
    (ref, callback) => {
        const onClickOutside = event => 
            ref.current && 
            ref.current.contains(event.target) === false && 
            callback(DEFAULT_ACTIVE_MENU_ITEM);

        useEffect(() => {
            document.addEventListener('click', onClickOutside);

            return () => { 
                document.removeEventListener('click', onClickOutside); 
            }
        });
    };

export const useGetDataFromRestApiEffect = 
    (url, setData) => useEffect(() => {
        fetch(url)
            .then(response => response.json())
            .then(normalizeData)
            .then(data => setData(data))
            .catch(error => console.log(error));
    });
