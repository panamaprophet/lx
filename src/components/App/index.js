import React, {useRef, useState} from 'react';

import Menu from '../Menu';

import styles from './styles.css';

import {DEFAULT_ACTIVE_MENU_ITEM, DATA_URL} from './constants';
import {useOutsideClickEffect, useGetDataFromRestApiEffect} from './hooks';

const App = () => {
    const ref = useRef(null);
    const [activeItemId, setActiveItemId] = useState(DEFAULT_ACTIVE_MENU_ITEM);
    const [menuItems, setMenuItems] = useState([]);

    useGetDataFromRestApiEffect(DATA_URL, setMenuItems);
    useOutsideClickEffect(ref, setActiveItemId);

    const toggleActiveItemId = id => 
        setActiveItemId(activeItemId === id ? DEFAULT_ACTIVE_MENU_ITEM : id);

    return (
        <div ref={ref} className={styles.root}>
            <Menu
                items={menuItems}
                activeItemId={activeItemId}
                onClick={toggleActiveItemId}
            />
        </div>
    );
};

export default App;