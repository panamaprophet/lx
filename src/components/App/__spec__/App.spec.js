import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';

import App from '../index';

describe('App', () => {
    test('component renders with default props', () => {
        const wrapper = shallow(<App />);
        expect(wrapper).toMatchSnapshot();
    });
});