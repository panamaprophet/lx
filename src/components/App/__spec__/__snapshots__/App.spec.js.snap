// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders with default props 1`] = `
<div
  className="root"
>
  <Menu
    activeItemId={-1}
    items={Array []}
    onClick={[Function]}
  />
</div>
`;
